File: tests/Tests.CSharp/DvKeyTests.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp;

namespace Zeta.Tests.CSharp;

public class DvKeyTests
{
    private static DvKey Row(params (string Key, DynamicValue Value)[] kvs)
    {
        var pairs = ImmutableArray.CreateBuilder<KeyValuePair<string, DynamicValue>>();
        foreach (var kv in kvs)
        {
            pairs.Add(new KeyValuePair<string, DynamicValue>(kv.Key, kv.Value));
        }
        return DvKey.OfValue(new DynamicValue.Object(pairs.ToImmutable()));
    }

    private static string RepoRoot()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(DvKeyTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
        {
            dir = dir.Parent;
        }
        return dir?.FullName ?? throw new DirectoryNotFoundException("Could not locate repo root (Zeta.sln).");
    }

    private static string Str(JsonElement el, string prop) =>
        el.GetProperty(prop).GetString()
            ?? throw new InvalidOperationException($"fixture property '{prop}' is not a string");

    private static DynamicValue BuildValue(JsonElement el)
    {
        string tag = Str(el, "t");
        switch (tag)
        {
            case "null":
                return new DynamicValue.Null();
            case "bool":
                return new DynamicValue.Bool(el.GetProperty("v").GetBoolean());
            case "int":
                return new DynamicValue.Int(long.Parse(Str(el, "v"), CultureInfo.InvariantCulture));
            case "float":
                return new DynamicValue.Float(
                    BitConverter.UInt64BitsToDouble(
                        ulong.Parse(Str(el, "v"), NumberStyles.HexNumber, CultureInfo.InvariantCulture)));
            case "str":
                return new DynamicValue.String(Str(el, "v"));
            case "bytes":
                return new DynamicValue.Bytes(Convert.FromHexString(Str(el, "v")).ToImmutableArray());
            case "arr":
                return new DynamicValue.Array(
                    el.GetProperty("v").EnumerateArray().Select(BuildValue).ToImmutableArray());
            case "obj":
                return new DynamicValue.Object(
                    el.GetProperty("v").EnumerateArray()
                        .Select(pair =>
                        {
                            JsonElement[] parts = pair.EnumerateArray().ToArray();
                            string key = parts[0].GetString() ?? "";
                            return new KeyValuePair<string, DynamicValue>(key, BuildValue(parts[1]));
                        })
                        .ToImmutableArray());
            default:
                throw new InvalidOperationException($"unsupported tag: {tag}");
        }
    }

    [Fact]
    public void EqualDynamicValueRowsGiveEqualKeysAndDistinctGiveDistinctKeys()
    {
        var a = Row(("id", new DynamicValue.Int(1L)), ("name", new DynamicValue.String("x")));
        var a2 = Row(("id", new DynamicValue.Int(1L)), ("name", new DynamicValue.String("x")));
        var b = Row(("id", new DynamicValue.Int(2L)), ("name", new DynamicValue.String("x")));

        Assert.Equal(a, a2);
        Assert.Equal(a.GetHashCode(), a2.GetHashCode());
        Assert.NotEqual(a, b);
    }

    [Fact]
    public void LexicographicalComparisonComparesCanonicalBytes()
    {
        var a = Row(("id", new DynamicValue.Int(1L)));
        var b = Row(("id", new DynamicValue.Int(2L)));

        Assert.True(a.CompareTo(b) < 0);
        Assert.True(b.CompareTo(a) > 0);
        Assert.Equal(0, a.CompareTo(a));
    }

    [Fact]
    public void CrossVerifyDvKeyVectorsMatchExpected()
    {
        var root = RepoRoot();
        var jsonPath = Path.Combine(root, "tests", "cross-verification", "dv-key-cloud-events", "vectors.json");
        Assert.True(File.Exists(jsonPath), $"vectors.json not found: {jsonPath}");

        using var doc = JsonDocument.Parse(File.ReadAllText(jsonPath));
        var vectors = doc.RootElement.GetProperty("dv_key_vectors").EnumerateArray();

        foreach (var v in vectors)
        {
            var val = BuildValue(v.GetProperty("value"));
            var expectedCborHex = Str(v, "expected_cbor_hex");
            var expectedHash = Str(v, "expected_hash");

            var key = DvKey.OfValue(val);
            var actualCborHex = Convert.ToHexString(key.Canonical).ToLowerInvariant();
            var actualHash = key.GetHashCode().ToString(CultureInfo.InvariantCulture);

            Assert.Equal(expectedCborHex, actualCborHex);
            Assert.Equal(expectedHash, actualHash);
        }
    }
}

File: tests/Tests.FSharp/DvKey.Tests.fs
module Zeta.Tests.DvKeyTests

open global.Xunit
open Zeta.Core
open System
open System.IO
open System.Text.Json
open System.Globalization
open System.Collections.Immutable

module CDC = Zeta.Core.DebeziumCdc

type private Marker = class end

let private repoRoot () : string =
    let assembly = typeof<Marker>.Assembly
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(assembly.Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then
        raise (InvalidOperationException("Could not locate repo root (Zeta.sln) from test assembly location."))
    dir.FullName

let rec private buildValue (el: JsonElement) : DynamicValue =
    let tag = el.GetProperty("t").GetString()
    match tag with
    | "null" -> DynamicValue.Null
    | "bool" -> DynamicValue.Bool(el.GetProperty("v").GetBoolean())
    | "int" -> DynamicValue.Int(int64 (el.GetProperty("v").GetString()))
    | "float" ->
        let bits = UInt64.Parse(el.GetProperty("v").GetString(), NumberStyles.HexNumber, CultureInfo.InvariantCulture)
        DynamicValue.Float(BitConverter.UInt64BitsToDouble(bits))
    | "str" -> DynamicValue.String(el.GetProperty("v").GetString())
    | "bytes" ->
        let hexStr = el.GetProperty("v").GetString()
        DynamicValue.Bytes(ImmutableArray.Create<byte>(Convert.FromHexString(hexStr)))
    | "arr" ->
        let arr = el.GetProperty("v").EnumerateArray() |> Seq.map buildValue |> Seq.toList
        DynamicValue.Array(arr)
    | "obj" ->
        let obj =
            el.GetProperty("v").EnumerateArray()
            |> Seq.map (fun pair ->
                let parts = pair.EnumerateArray() |> Seq.toArray
                parts.[0].GetString(), buildValue parts.[1]
            )
            |> Seq.toList
        DynamicValue.Object(obj)
    | _ -> failwithf "unsupported tag: %s" tag

let private row (kvs: (string * DynamicValue) list) = DvKey.ofValue (DynamicValue.Object kvs)

[<Fact>]
let ``equal DynamicValue rows give equal keys; distinct give distinct keys`` () =
    let a = row [ "id", DynamicValue.Int 1L; "name", DynamicValue.String "x" ]
    let a2 = row [ "id", DynamicValue.Int 1L; "name", DynamicValue.String "x" ]
    let b = row [ "id", DynamicValue.Int 2L; "name", DynamicValue.String "x" ]
    Assert.Equal<DvKey>(a, a2)
    Assert.Equal(a.GetHashCode(), a2.GetHashCode())
    Assert.NotEqual<DvKey>(a, b)

[<Fact>]
let ``DvKey rows work as ZSet keys (DynamicValue rows in a Z-set)`` () =
    let z = ZSet.ofSeq [ row [ "id", DynamicValue.Int 1L ], 1L; row [ "id", DynamicValue.Int 2L ], 1L ]
    Assert.Equal(2, z.Count)
    Assert.Equal(1L, ZSet.lookup (row [ "id", DynamicValue.Int 1L ]) z)

[<Fact>]
let ``Debezium change events over DynamicValue rows convert to Z-set deltas (end-to-end)`` () =
    let before = row [ "id", DynamicValue.Int 1L; "v", DynamicValue.String "old" ]
    let after = row [ "id", DynamicValue.Int 1L; "v", DynamicValue.String "new" ]
    // an update of a DynamicValue row = retract old + insert new
    Assert.Equal<ZSet<DvKey>>(ZSet.ofSeq [ before, -1L; after, 1L ], CDC.toZSetDelta (CDC.update before after))
    // and the delta round-trips at the delta level
    let delta = CDC.toZSetDelta (CDC.create after)
    let rt = CDC.ofZSetDelta delta |> List.map CDC.toZSetDelta |> List.fold (+) ZSet.Empty
    Assert.Equal<ZSet<DvKey>>(delta, rt)

[<Fact>]
let ``CrossVerifyDvKeyVectorsMatchExpected`` () =
    let root = repoRoot ()
    let jsonPath = Path.Combine(root, "tests", "cross-verification", "dv-key-cloud-events", "vectors.json")
    Assert.True(File.Exists(jsonPath), sprintf "vectors.json not found: %s" jsonPath)

    use doc = JsonDocument.Parse(File.ReadAllText(jsonPath))
    let vectors = doc.RootElement.GetProperty("dv_key_vectors").EnumerateArray()

    for v in vectors do
        let expectedCborHex = v.GetProperty("expected_cbor_hex").GetString()
        let expectedHash = v.GetProperty("expected_hash").GetString()
        let valEl = v.GetProperty("value")
        let value = buildValue valEl

        let key = DvKey.ofValue value
        let actualCborHex = Convert.ToHexString(DvKey.canonical key).ToLowerInvariant()
        let actualHash = string (key.GetHashCode())

        Assert.Equal(expectedCborHex, actualCborHex)
        Assert.Equal(expectedHash, actualHash)
